### PR TITLE
ci: revert file changes and add some checks

### PR DESCRIPTION
**Pull request by lgfa29**
During the release there are several files that need to be modified:

  - .release/ci.hcl: the notification channel needs to be updated to a
    channel with greater team visibility during the release.
  - version/version.go: the Version and VersionPrerelease variables
    need to be set so they match the release version.

After the release these files need to be reverted.

For GA releases the following additional changes also need to happen:

  - version/version.go: the Version variable needs to be bumped to the
    next version number.
  - GNUMakefile: the LAST_RELEASE variable needs to be set to the
    version that was just released.

Since the release process will commit file changes to the branch being
used for the release, it should _never_ run on main, so the first step
is now to protect against that.

It also adds a validation to make the user input version is correct.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -122,3 +149,41 @@ jobs:
           token: ${{ secrets.ELEVATED_GITHUB_TOKEN}}
           inputs: '{"build-ref": "${{ steps.commit-change-push.outputs.build-ref }}", "make-prerelease": "false"}'
           ref: ${{ needs.prepare-release.outputs.build-ref }}
+
+      - name: Revert notification channel
+        if: ${{ github.event.inputs.notification-channel != '' }}
+        run: |
+          sed -i.bak -e 's|\(notification_channel * = *"\)[^"]*|\1${{ steps.notification-channel.outputs.prev-channel }}|g' .release/ci.hcl
+          rm -rf .release/ci.hcl.bak
+          git diff --color=always .release/ci.hcl
```
Comment (tgross) [inline .github/workflows/release.yml:158]: We're calling this "revert" but it's not obvious whether this literally backing out "Generate files for..." commit or making a slightly different change that could potentially drift. Could we literally reset this file with git (ex. `git reset $ref -- .release/ci.hcl`)?

This is particularly the case for the step below too, where we're doing two different things.

**Review by tgross — APPROVED** [review]
LGTM

File: .github/workflows/release.yml (hunk 1 above)
Comment (lgfa29) [reply .github/workflows/release.yml:158]: Ah yeah, good point. This one is direct revert, the next one not quite, so I will rename it as well. Thanks!

**Review by DerekStrickland — COMMENTED** [review]
Apologies in advance if I'm asking a question that doesn't make sense.

Since we have conditional logic based on branch (I think?) could we use an approach similar to what is discussed [here](https://stackoverflow.com/questions/58139406/only-run-job-on-specific-branch-with-github-actions) so that we don't have to manually massage files?  I'm sure you already considered something like this. I'm just trying to learn more about the context / challenge.

**Review by shore — APPROVED** [review]

**Comment by lgfa29** [comment]
> Apologies in advance if I'm asking a question that doesn't make sense.
> 
> Since we have conditional logic based on branch (I think?) could we use an approach similar to what is discussed [here](https://stackoverflow.com/questions/58139406/only-run-job-on-specific-branch-with-github-actions) so that we don't have to manually massage files? I'm sure you already considered something like this. I'm just trying to learn more about the context / challenge.

I didn't quite follow that. The conditional logic is based on the prerelease part of the version being release. 

The only branch check we have is to prevent it from running from `main`. I could extract that first step into a separate and have it as a dependency to the main one, but each job is a separate environment in GitHub Actions, which means it would have to spin a whole new infrastructure just for a `return 1` 😅 

Is this what you had in mind?

**Pull request merged**